### PR TITLE
feat: guarantee correct URL

### PR DESCRIPTION
**Pull request by mLuca**
Serialization/normalization guarantees valid URI values according to JSON/XML specification

fixes https://github.com/CycloneDX/cyclonedx-javascript-library/issues/992

### CODE REVIEW

**Comment by mLuca** [comment]
I am not a Typescript developer, yet I want to have this fix -> I propose a concept for fixing this issue in order to get feedback

File: src/serialize/json/normalize.ts [hunk 1]
```diff
@@ -446,20 +446,26 @@ export class LicenseNormalizer extends BaseJsonNormalizer<Models.License> {
...
         url: JsonSchema.isIriReference(url)
-          ? url
+          ? encodeURI(url)
           : undefined
       }
     }
   }
 
   #normalizeSpdxLicense (data: Models.SpdxLicense, options: NormalizerOptions): Normalized.SpdxLicense {
+    let encodedUrl: string | undefined
```
Comment (jkowalleck) [inline src/serialize/json/normalize.ts:456]: simplify:
```ts
data.url === undefned ? 
  ? undefined
  : encodeURI(data.url.toString())  
```

**Review by jkowalleck — CHANGES_REQUESTED** [review]

**Comment by jkowalleck** [comment]
the proposed solution is insufficient. it will double-encode.
test with an already properly encoded url: 
```js
encodeURI("https://acme.org/foo%5B%5D=bar")
// actual: "https://acme.org/foo%255B%255D=bar"
// expected: "https://acme.org/foo%5B%5D=bar"
```

File: src/serialize/json/normalize.ts [hunk 2]
```diff
@@ -446,7 +447,7 @@ export class LicenseNormalizer extends BaseJsonNormalizer<Models.License> {
           ? undefined
           : this._factory.makeForAttachment().normalize(data.text, options),
         url: JsonSchema.isIriReference(url)
-          ? url
+          ? percentEncodeUrl(url)
```
Comment (jkowalleck) [inline src/serialize/json/normalize.ts:450]: the encoding should have happened before the `JsonSchema.isIriReference()` is run on it.

File: src/serialize/json/normalize.ts [hunk 3]
```diff
@@ -459,7 +460,9 @@ export class LicenseNormalizer extends BaseJsonNormalizer<Models.License> {
         text: data.text === undefined
           ? undefined
           : this._factory.makeForAttachment().normalize(data.text, options),
-        url: data.url?.toString()
+        url: data.url === undefined
```
Comment (jkowalleck) [inline src/serialize/json/normalize.ts:463]: I forgot to add the `JsonSchema.isIriReference()` here.
could you add it?
if it returned false, the the url should be omitted

File: src/serialize/json/normalize.ts [hunk 4]
```diff
@@ -514,7 +517,7 @@ export class ExternalReferenceNormalizer extends BaseJsonNormalizer<Models.Exter
   normalize (data: Models.ExternalReference, options: NormalizerOptions): Normalized.ExternalReference | undefined {
     return this._factory.spec.supportsExternalReferenceType(data.type)
```
Comment (jkowalleck) [inline src/serialize/json/normalize.ts:518]: the encoding should have happened before the `JsonSchema.isIriReference()` is run on it?
if it returns false, then `normalize` should return `undefined`

File: src/serialize/xml/normalize.ts [hunk 5]
```diff
@@ -564,7 +565,7 @@ export class LicenseNormalizer extends BaseXmlNormalizer<Models.License> {
           ? undefined
           : this._factory.makeForAttachment().normalize(data.text, options, 'text'),
         XmlSchema.isAnyURI(url)
-          ? makeTextElement(url, 'url')
+          ? makeTextElement(percentEncodeUrl(url), 'url')
```
Comment (jkowalleck) [inline src/serialize/xml/normalize.ts:568]: `percentEncodeUrl()`  should be applied before `XmlSchema.isAnyURI(url)` is checked

File: src/serialize/xml/normalize.ts [hunk 6]
```diff
@@ -641,7 +642,7 @@ export class SWIDNormalizer extends BaseXmlNormalizer<Models.SWID> {
 
 export class ExternalReferenceNormalizer extends BaseXmlNormalizer<Models.ExternalReference> {
   normalize (data: Models.ExternalReference, options: NormalizerOptions, elementName: string): SimpleXml.Element | undefined {
```
Comment (jkowalleck) [inline src/serialize/xml/normalize.ts:644]: the encoding should have happened before the `XmlSchema.isAnyURI()` is run on it?
if it returns false, then `normalize` should return `undefined`

File: src/_helpers/percentEncodedUrl.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+SPDX-License-Identifier: Apache-2.0
+Copyright (c) OWASP Foundation. All Rights Reserved.
+*/
+
+const encoderMap = new Map<string, string>([
```
Comment (jkowalleck) [inline src/_helpers/percentEncodedUrl.ts:20]: a frozen dictionary would fit better, here

File: src/_helpers/percentEncodedUrl.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+  [' ', '%20'],
+  ['[', '%5B'],
+  [']', '%5D'],
+  ['<', '%3C'],
+  ['>', '%3E'],
+  ['{', '%7B'],
+  ['}', '%7D']
+])
+
+const regEx = new RegExp(Object.keys(encoderMap).join('|'), 'gi')
```
Comment (jkowalleck) [inline src/_helpers/percentEncodedUrl.ts:30]: why the `gi` flags? 
please write unit tests for `percentEncodeUrl` and see if the flags are needed,

**Comment by jkowalleck** [comment]
@mLuca i will add unit tests and fix your implementation.

File: src/serialize/xml/normalize.ts (hunk 6 above)
Comment (jkowalleck) [reply src/serialize/xml/normalize.ts:646]: done via https://github.com/CycloneDX/cyclonedx-javascript-library/pull/996/commits/95fb7c92e28f9bf5ed8f2378ad550f41d2f4d0bf

File: src/serialize/json/normalize.ts (hunk 4 above)
Comment (jkowalleck) [reply src/serialize/json/normalize.ts:521]: done via https://github.com/CycloneDX/cyclonedx-javascript-library/pull/996/commits/95fb7c92e28f9bf5ed8f2378ad550f41d2f4d0bf

File: src/serialize/json/normalize.ts (hunk 3 above)
Comment (jkowalleck) [reply src/serialize/json/normalize.ts]: done via https://github.com/CycloneDX/cyclonedx-javascript-library/pull/996/commits/95fb7c92e28f9bf5ed8f2378ad550f41d2f4d0bf

**Review by jkowalleck — APPROVED** [review]

**Pull request merged**